### ISSUE TITLE
Intermittent listener close issue popping up again

### ISSUE THREAD

**Issue by upalsaha** [riverqueue/river#256]
We definitely were getting hit with the previous hot loop issue mentioned in issue [#248](https://github.com/riverqueue/river/issues/248). After upgrading to `0.0.25` we did see the hot loop issue resolved with PR [#246](https://github.com/riverqueue/river/pull/246), but it seems ~15 minutes or so we're still getting the `failed to send closeNotify alert` issue sporadically. While the listener is able to reconnect after a while, was wondering if anyone had any updates on a resolution.

Example log:
<img width="744" alt="Screenshot 2024-03-08 at 2 28 10 PM" src="https://github.com/riverqueue/river/assets/6760014/27f8ac9b-a8a7-4ca5-bffc-ce812df44b24">

**Comment by brandur** [comment]
Thanks for the report @upalsaha.

Ugh, unfortunately given that seems to be a driver-level error, it's a little difficult to even see exactly where that's coming from. Beyond the error, are you noticing River being degraded at all? The notifier has a built in loop that's supposed to reconnect itself even on error.

It's not released yet, but I recently did a big refactor in the notifier from top to bottom as part of #253. Can't guarantee it solves this problem, but it should be overall more stable, and at the very least will give us more specific log messages to help track down the exact location of problems.

**Comment by upalsaha** [comment]
Hey @brandur appreciate the additional context. After your original fix River behavior isn't degraded and only noticeable issue is those logs popping up (with the reconnect thankfully working). Will give the new update a go and follow up!

**Comment by brandur** [comment]
We cut v0.1.0 with the changes.

**Comment by gabrielclimb** [comment]
should this issue be closed?

**Comment by brandur** [comment]
Probably. @upalsaha if you see any other trouble with this, okay to reopen and discuss, but closing for now.

**Comment by vblcc** [comment]
Hello again @brandur ,

as you probably already know, this is a tricky one to catch or reproduce.

However, I've noticed that pgxpool ignores this kind of "close error": https://github.com/jackc/pgx/blob/master/pgxpool/pool.go#L315 .
Best I can tell the libpq library does the same (pqClosePGconn and pqDropConnection don't return errors).

Would you be open to ignore this kind of error in River?
There is a proposal in the Go issues tracker to make closeNotify a dedicated error type but it hasn't received much attention: https://github.com/golang/go/issues/75600 .

**Comment by brandur** [comment]
@vblcc Is that Pgx link right? I'm just looking and I don't see any errors being ignored in that particular location.

**Comment by vblcc** [comment]
My bad, here is a link to the exact line (321) I'm referring to:
https://github.com/jackc/pgx/blob/master/pgxpool/pool.go#L321

The return from `conn.Close(ctx)` is silently being dropped, not "explicitly ignored" with an `_`, sorry for the confusion.

I tried to go back in pgx history to find out why but did not find any relevant commit explaining it.

**Comment by brandur** [comment]
@vblcc Ah, I see what you mean. Okay, just to circle back all the way around and make this as explicit as possible for me — is there a particular place in River code where you were thinking that errors should be ignored? Somewhere in the notifier?

**Comment by vblcc** [comment]
@brandur The error log we're seeing is `river: notifier.Notifier: Error closing listener` with the error
```
tls: failed to send closeNotify alert (but connection was closed anyway): write tcp 10.0.5.131:40192->10.0.7.96:5432: i/o timeout
```
which I believe comes from https://github.com/riverqueue/river/blob/master/internal/notifier/notifier.go#L266 . 

I'm thinking of the call to`pgx.Conn.Close()` down the call chain from there:
https://github.com/riverqueue/river/blob/master/riverdriver/riverpgxv5/river_pgx_v5_driver.go#L1046

My understanding is that `pgx` and `libpq` intentionally ignore errors when closing that connection. But I don't have any concrete documentation to back that up, so I understand that ignoring it is a big ask. 

Tuning the log level (either defaults or by configuration) could be an alternative. Or if this is an (understandable) wont-fix then I'll just have to try to get at it with log filtering.

**Comment by brandur** [comment]
@vblcc Awesome thanks for the great detail here. One more question: is it always that same error that you're seeing over and over again? (Relating to "closeNotify")

@bgentry Any strong opinions on this before I take a crack at it? My gut reaction here would be to special case common errors and downgrade their log level to debug, which should cover 99% of peoples needs I would think since it'd be implausibly noisy to run a prod instance on debug already.

**Comment by vblcc** [comment]
@brandur That is the only error I can see from the notifier or related to closeNotify.
If you mean errors generally then there is also the occasional:
```
msg: river: maintenance.PeriodicJobEnqueuer: Error starting transaction
err: context canceled
```

**Comment by brandur** [comment]
@vblcc Okay thanks! Hm, interesting with that cancellation on the job enqueuer (weird to see context cancelled that early), but definitely a separate issue.

**Comment by brandur** [comment]
Opened #1216.

**Comment by bgentry** [comment]
No objections, your solution looks good. Let's give it a shot!

**Comment by brandur** [comment]
Great!

Okay, shipped this out in v0.35.0.

**Comment by vblcc** [comment]
Hi @brandur 

The error popped up in our logs again. I think the "if" might be flipped?

**Comment by brandur** [comment]
Oops, yes my fault. Opened a fix in #1231.

**Comment by brandur** [comment]
Released v0.35.1.
